### PR TITLE
Fix branch creation at a point other than end-of-WAL

### PR DESCRIPTION
**Pull request by hlinnaka**
See commit messages for details

### CODE REVIEW

File: control_plane/src/local_env.rs [hunk 1]
```diff
@@ -325,19 +315,67 @@ pub fn create_branch(
...
+    let last_segno = upto.segment_number(wal_seg_size);
+    let last_segoff = upto.segment_offset(wal_seg_size);
+
+    for entry in fs::read_dir(src_dir).unwrap() {
+        if let Ok(entry) = entry {
+            let entry_name = entry.file_name();
+            let fname = entry_name.to_str().unwrap();
+
+            // Check if the filename looks like an xlog file, or a .partial file.
+            if !xlog_utils::IsXLogFileName(fname) && !xlog_utils::IsPartialXLogFileName(fname) {
```
Comment (ericseppanen) [inline control_plane/src/local_env.rs:345]: Seems like IsXlogFileName should accept a `&Path` or some `P: AsRef<Path>`...

File: test_runner/fixtures/zenith_fixtures.py [hunk 2]
```diff
@@ -126,29 +128,28 @@ def pageserver(zenith_cli):
...
+        self.zenith_cli.run(['pg', 'start', 'pg{}'.format(self.instance_num)])
         self.running = True
-
+        return
 
     #lines should be an array of valid postgresql.conf rows
     def config(self, lines):
-        #TODO use real node name, not just guessed pg1
-        filename = 'pgdatadirs/pg1/postgresql.conf'
+        filename = 'pgdatadirs/pg{}/postgresql.conf'.format(self.instance_num)
```
Comment (ericseppanen) [inline test_runner/fixtures/zenith_fixtures.py:152]: I wish we didn't have to guess at the name "pgX" that was created.

Should we get this as feedback from the CLI? Should we specify the name we want?

File: test_runner/fixtures/zenith_fixtures.py [hunk 3]
```diff
@@ -126,29 +128,28 @@ def pageserver(zenith_cli):
...
+    def __init__(self, zenith_cli, repo_dir, instance_num):
         self.zenith_cli = zenith_cli
+        self.instance_num = instance_num
         self.running = False
         self.host = 'localhost'
-        self.port = 55432
+        self.port = 55431 + instance_num
         self.repo_dir = repo_dir
-        # path to conf is <repo_dir>/pgdatadirs/pg1/postgresql.conf
+        # path to conf is <repo_dir>/pgdatadirs/pg<instance_num>/postgresql.conf
```
Comment (ericseppanen) [inline test_runner/fixtures/zenith_fixtures.py:138]: I think this comment was originally just a note to myself. It doesn't seem like we need it here so feel free to remove it.

File: zenith/src/main.rs [hunk 4]
```diff
@@ -347,7 +339,7 @@ fn parse_point_in_time(s: &str) -> Result<local_env::PointInTime> {
 
         let mut result = parse_point_in_time(&pointstr)?;
 
-        result.lsn = lsn.unwrap_or(0);
+        result.lsn = lsn.unwrap_or(Lsn(0));
```
Comment (ericseppanen) [inline zenith/src/main.rs:342]: I'm curious why the result might be `Lsn(0)`?  Should this be `Option<Lsn>`?

**Review by ericseppanen — APPROVED** [review]
Nice!

How did you find the python framework?

File: test_runner/fixtures/zenith_fixtures.py (hunk 2 above)
Comment (ericseppanen) [reply test_runner/fixtures/zenith_fixtures.py:152]: And by "I wish" I don't mean "should fix now". I just meant "at some point in the future".

File: control_plane/src/local_env.rs (hunk 1 above)
Comment (hlinnaka) [reply control_plane/src/local_env.rs:345]: Hmm, I don't know, the argument is just a filename, not a path. For comparison, `File::file_name(&self)` also returns a string, not a path. To be precise, it returns an `&OsStr`. Maybe IsXLogFileName() should also accept `&OsStr`?

File: zenith/src/main.rs (hunk 4 above)
Comment (hlinnaka) [reply zenith/src/main.rs:342]: Yeah, that probably would be more idiomatic Rust than using 0 as a special value.

**Comment by hlinnaka** [comment]
> How did you find the python framework?

I was pretty frustrated at first, to be honest. The existing fixture didn't really support what I was trying to test, that is, creating branches and running multiple postgres instances on them. So I had to dig into python documentation on how fixtures work, that was new to me, too. But after I managed to add the features I needed to the fixture - and I'm pretty happy with the way that turned out in the end - writing the test was quite alright and the test runner gave pretty good feedback on failures. Writing the next tests should be easier.

**Comment by hlinnaka** [comment]
Oh wait, CircleCI isn't running the new test. Do I have to add it to `.circleci/config.yml` as a new job alongside the test_pgbench.py and test_pg_regress.py tests? Seems a bit heavy weight to have a separate job for it

**Comment by hlinnaka** [comment]
> Oh wait, CircleCI isn't running the new test. Do I have to add it to `.circleci/config.yml` as a new job alongside the test_pgbench.py and test_pg_regress.py tests? Seems a bit heavy weight to have a separate job for it

I think the `test_config.py` and `test_pageserver_api.py` that @alubennikova and @patins have addd recently are also not being run. Clearly we need a better system, if all three of us independently missed that :-). Can we have one CircleCI job to run all the tests in the directory?

**Comment by kelvich** [comment]
> I think the `test_config.py` and `test_pageserver_api.py` that @alubennikova and @patins have addd recently are also not being run. Clearly we need a better system, if all three of us independently missed that :-). Can we have one CircleCI job to run all the tests in the directory?

@funbringer is there any good option to add automatic test discovery?

File: control_plane/src/local_env.rs (hunk 1 above)
Comment (ericseppanen) [reply control_plane/src/local_env.rs:345]: Sure, `AsRef<OsStr>` is even more flexible. That way the argument [could be any of](https://doc.rust-lang.org/std/ffi/struct.OsStr.html#impl-AsRef%3COsStr%3E): `OsString`, `OsStr`, `String`, `str`, `Path`, `PathBuf`, etc.

**Comment by ericseppanen** [comment]
> Oh wait, CircleCI isn't running the new test. Do I have to add it to `.circleci/config.yml` as a new job alongside the test_pgbench.py and test_pg_regress.py tests? Seems a bit heavy weight to have a separate job for it

> I think the `test_config.py` and `test_pageserver_api.py` that @alubennikova and @patins have addd recently are also not being run. Clearly we need a better system, if all three of us independently missed that :-). Can we have one CircleCI job to run all the tests in the directory?

I was delaying making a decision until we see what the tests look like; we can set it up any way we like.  My instinct is to lump similar tests together in a small number of python files, and then run one CircleCI job per file.

There are many other options, though. Pytest can select tests by file, collection of files, directory, test name, pattern matching on test names...

**Comment by ericseppanen** [comment]
There is no way (that I've found) to create new CircleCI jobs without modifying `.circleci/config.yml`.

My hope was that we could keep the number of python files small, and mostly just add new tests to the already-existing files. And then when a test starts taking too long to run (> 10 minutes?) break that file in two, and add another circleci job.

**Comment by funbringer** [comment]
> @funbringer is there any good option to add automatic test discovery?

Sure, we could use [pytest](https://docs.pytest.org/en/6.2.x/goodpractices.html#conventions-for-python-test-discovery) or something like that.

**Comment by kelvich** [comment]
Let's commit then? It would be handy if I can rebase my current work on top of this -- there is a lot of changes to the same files.

**Review by kelvich — APPROVED** [review]

**Comment by kelvich** [comment]
On the test matter: we can have whitelist of tests that go to a separate jobs (as it is coded right now), and one job that runs rest of tests (if any)

**Comment by ericseppanen** [comment]
> On the test matter: we can have whitelist of tests that go to a separate jobs (as it is coded right now), and one job that runs rest of tests (if any)

It will be easiest if we work within the filtering capabilities that pytest has built-in. For example, the easiest way to have a big pile of tests run sequentially within a CircleCI job is to place them all within one "batch" directory.

I propose something like this:
```text
test_runner/
    fixtures/
    batch_a/
        foo.py
        bar.py
    standalone/
        baz.py
```

We could then have a `batch_a` CircleCI job that runs `pytest batch_a`, and pytest will scan that directory and run all the tests in `foo.py` and `bar.py`. If the batch grows too big, we can add `batch_b`, etc.

We would then manually configure CircleCI jobs for each standalone test: e.g. a `baz` job that runs `pytest standalone/baz.py`.

**Comment by hlinnaka** [comment]
> On the test matter: we can have whitelist of tests that go to a separate jobs (as it is coded right now), and one job that runs rest of tests (if any)

Ok, I'm going to push this as it is. And we'll figure out how we want to add all the missing tests to CircleCI separately.

**Comment by hlinnaka** [comment]
I opened https://github.com/zenithdb/zenith/pull/141 to rearrange the tests so that CircleCI runs them. And a bunch of other test refactoring. Please take a look

**Pull request merged**